### PR TITLE
[AMDGPU] Fix lack of LDS DMA check in the AA handling

### PR DESCRIPTION
**Pull request by rampitec**
SIInstrInfo::areMemAccessesTriviallyDisjoint does a DS offset checks, but does not account for LDS DMA instructions. Added these checks. Without it code falls through and returns true which is wrong. As a result mayAlias would always return false for LDS DMA and a regular LDS instruction or 2 LDS DMA instructions.

At the moment this is NFCI because we do not use this AA in a context which may touch LDS DMA instructions. This is also unreacheable now because of the ordered memory ref checks just above in the function and LDS DMA is marked as volatile. This volatile marking is removed in PR #75247, therefore I'd submit this check before #75247.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 1]
```diff
@@ -3656,8 +3656,8 @@ bool SIInstrInfo::areMemAccessesTriviallyDisjoint(const MachineInstr &MIa,
   // underlying address space, even if it was lowered to a different one,
   // e.g. private accesses lowered to use MUBUF instructions on a scratch
   // buffer.
-  if (isDS(MIa)) {
-    if (isDS(MIb))
+  if (isDS(MIa) || isLDSDMA(MIa)) {
+    if (isDS(MIb) || isLDSDMA(MIb))
       return checkInstOffsetsDoNotOverlap(MIa, MIb);
```
[the remark this answers is not in the document]
Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:3661]: It does, even though it just bails. It goes down to getMemOperandsWithOffsetWidth and there it bails on the LDS DMA:
```
 449│     // Get appropriate operand, and compute width accordingly.
 450│     DataOpIdx = AMDGPU::getNamedOperandIdx(Opc, AMDGPU::OpName::vdst);
 451│     if (DataOpIdx == -1)
 452│       DataOpIdx = AMDGPU::getNamedOperandIdx(Opc, AMDGPU::OpName::vdata);
 453│     if (DataOpIdx == -1) // LDS DMA
 454│       return false;
```
In principle these offsets are analyzable. This is a typical store memop:
```
(dereferenceable store (s32) into `ptr addrspace(3) getelementptr inbounds (%llvm.amdgcn.kernel.buffer_load_lds_dword_2_ar
rays.lds.t, ptr addrspace(3) @llvm.amdgcn.kernel.buffer_load_lds_dword_2_arrays.lds, i32 0, i32 1)
```
But is you want I can bail right here.

[the remark this answers is not in the document]
Comment (rampitec) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp]: Just bail early.

**Comment by rampitec** [comment]
Ping. This one seems obvious to me.

**Review by searlmc1 — APPROVED** [review]
LGTM

**Pull request merged**